### PR TITLE
use a common render function to write all the prompts

### PR DESCRIPTION
**Pull request by coryb**
Hey @AlecAivazis, the goal of this PR is to isolate the prompt rendering and subsequent refresh and cursor movement into a common routine.  I did this by embedding a common `renderer` object into each prompt.  I couldn't find another way that made any sense. Overall I think it is a simplification although `Password` became more complicated due to the way `readline` works.  `readline` really wants to manage the prompt and tries to do a bunch of tricky cursor manipulation with no obvious ways to work around it.  I spent a lot of time fighting with `readline`.

This PR also merges the `Question` and `Choice` templates into a single template so that we can have common render logic for `Select` and `MultiSelect`.

I migrated the tests to be more table driven, which simplifies them, I also started using `github.com/stretchr/testify/assert` which I have found to be very useful.

Lastly I fixed a minor bug with the HideCursor, where if you ctrl-c the prompt the ShowCursor was not getting called.  So I wrapped that in a defer to always call ShowCursor even if interrupted.

Anyway, check it out and let me know if you have any questions or concerns.  Thanks!

### CODE REVIEW

**Comment by AlecAivazis** [comment]
Hey @coryb thanks for yet another PR. I will try to make my way through this tomorrow, i'm having a bit of trouble focusing tracking the changes related to a single one of these features. I'm not sure how tightly coupled they are but if it would be possible to split this into a few different PRs that would be a huge help. 

I want to make sure we can get your stuff in as soon as possible and don't want a large PR to stall the rest of the improvements. My first reaction is that the move to test based tables is a huge improvement, i like the unification of the rendering-refresh logic but I haven't gotten a chance to grok the password changes. 

Thanks again!

**Comment by coryb** [comment]
Hey, yeah, sorry it is another big PR, I just worked my way through all the various issues I ran into and it was bigger than I hoped. Most of it is a bit coupled, so I am not sure how I can pull them apart.  I can submit one for the HideCursor bit, that is stand-alone.  I can also submit one for just one prompt type (perhaps `Input` with the renderer logic), plus the testing refactor for that prompt.  I can leave the other prompt types out for now to keep it simple.

**Comment by AlecAivazis** [comment]
well i suppose I see three separate parts: the `HideCursor` fix, the refactor of the tests to tables, and the renderer refactor. I'm totally fine with having all of the inputs as one, just want to try to think about this renderer addition without worrying about holding back the other improvements

**Comment by coryb** [comment]
okay, I will see what I can do.

**Comment by coryb** [comment]
Okay, I have created #41 and #42 for the `HideCursor` issue and table driven tests.  Assuming they are merged to master I will then merge them into this PR.

**Comment by AlecAivazis** [comment]
I just saw, thanks so much for doing that! Make it much easier for me to merge them with confidence. I will review the two tonight assuming I dont fall asleep soon and tackle the third tomorrow

**Comment by coryb** [comment]
Okay, cool, I am going to pass out myself, so I will update this PR in the morning if the others get merged.  Thanks!

File: terminal/print.go [hunk 1]
```diff
@@ -5,21 +5,21 @@ import (
 )
 
 var (
-	ansiStdout = NewAnsiStdout()
+	AnsiStdout = NewAnsiStdout()
```
Comment (coryb) [inline terminal/print.go]: fyi, made this public so I could override it to capture output for testing.

**Comment by coryb** [comment]
Okay, updated the PR with the merge from master.  That should should help reduce some of the noise.

**Comment by coryb** [comment]
btw, I noticed you removed most of the examples, just wondering why.  I found the extra examples useful for testing changes like these, so I wrote them all again, but I have not checked them in.

**Comment by AlecAivazis** [comment]
@coryb sorry it's taken me so long to go through this PR. I'm in the final stretch before a big release at work so all of my focus is there. That being said, I'm starting to go through it now. 

I moved your examples to the test directory which is where I have been putting all of the integration tests and edge cases that have been reported. Is there one missing?

**Comment by coryb** [comment]
Hey, no problem, work is useful for paying the bills.  

Doh, I missed the examples in the tests directory, that explains it :)

**Comment by AlecAivazis** [comment]
Haha no worries, I haven't really found a really satisfying way to prevent regressions as we work on this package. Keeping a directory of all of the different issues and examples people have provided was the simplest thing for now. 

 I meant for the example dir to be more consumable snippets of code for common use cases. Although I'm not sure it's really that useful now that I've fleshed out the README

**Comment by coryb** [comment]
I suppose we would use `expect` to automate prompt and response testing in a quasi automated fashion.  If I get some free time I will play with it.

**Comment by AlecAivazis** [comment]
Wow, i hadn't heard of `expect` before, this looks really promising! and https://github.com/ThomasRooney/gexpect looks like a great way to keep external dependencies to a minimum

File: password.go [hunk 2]
```diff
@@ -17,9 +20,49 @@ var PasswordQuestionTemplate = `
...
+// does not play nice here.  Write is called with clear-line statements which need to be
+// ignored then it is called with the full buffered output. So if you type "hi" this is the
+// sequence that will be sent to Write
+// 0: "\033[J\033[2K\r"
+// 1: " \b"
+// 2: "\033[J\033[2K\r"
+// 3: "*" (this is the masked "h" from "hi")
+// 4: "\033[J\033[2K\r"
+// 5: "**" (this is the masked "hi" from "hi")
+func (o *filteredReadlinePasswordOutput) Write(p []byte) (n int, err error) {
```
Comment (AlecAivazis) [inline password.go]: Hmm so I still have to go through this a bit more but I wonder if it's better to just leave password as sort of an odd man and remove the `renderer` from its implementation. There is a massive win in unifying all of the re-rendering logic but I fear this might cause more bugs than the pain in having to potentially change two different places for rendering logic.

File: terminal/print.go (hunk 1 above)
Comment (AlecAivazis) [inline terminal/print.go]: Yea i noticed, nice idea - since the `Ansi` bit is a bit of an artifact of the old package, how do you think about just naming this `Stdout`? I know I probably make some dumb naming request on every PR (sorry!) but the API is already a bit verbose, i think it would be good to try to minimize it wherever possible

File: password.go (hunk 2 above)
Comment (coryb) [inline password.go]: I understand your concern, we could do that.   The main reason for this refactor for me is to be able to generate `help` messages in the prompt directly.  So my current plan is to have something like a `Help string` property for each Prompt which will allow me to provide hints to prompts like:
``` 
    prompt := &survey.Input{
        Message: "Enter birthday:",
        Help: "Birthday format must be in Day/Month/Year",
    }
```
Then the prompt would look something like this by default:
```
? Enter birthday: [? for help] 
```
then if user enters `?` we rewrite the prompt as:
```
Birthday format must be in Day/Month/Year
? Enter birthday: 
```
So if we make Password the odd-one-out it would also not be able to used to participate in the Help messages.  I would more confident about leaving it out if we had a strategy/plan to pull it back in but short of purging readline, not sure what else we could do.

**Comment by coryb** [comment]
I think `gexpect` is cheating, you need to get familiar with the up-and-coming language `tcl` with expect :)

File: terminal/print.go (hunk 1 above)
Comment (coryb) [inline terminal/print.go]: okay, np, the naming is pretty arbitrary to me.  Really we should probably replace all the `fmt.Print*(...)` calls with `fmt.Fprint*(Stdout, ...)` so we can test/capture all the output, but that is probably a change for later.

Comment (coryb) [inline terminal/print.go]: updated

File: password.go (hunk 2 above)
Comment (AlecAivazis) [inline password.go]: There you go again coming up with ideas that make me question the interface. 

I really like this and it feels like something that a question should be concerned with instead of a prompt but i don't really like the idea of passing `helpText` to `Prompt` although it would closer resemble the signature of `Cleanup`. 

Is this something you imagine on multi-selects too? I'm not sure how one would collect a request for help in the various multi-line prompts. In the case of non-`Input` prompts, i'm not sure how I feel about showing help text based on an input the user cannot see they typed. FWIW, Inquirer somewhat addresses in multi-line prompts by showing a short help text describing the arrow key behaviour until you press one.

Comment (AlecAivazis) [inline password.go]: Also, i'm not opposed to removing readline - i did it to remove the old logic I had in this package to listen to `/dev/tty` which always made me uneasy even though windows seemed to support it. So long as we can keep cross platform support, anything is on the table if it improves the package.

Comment (AlecAivazis) [inline password.go]: Second also, maybe it's worth removing the `renderer` from `Password` and we can try to pull that logic into somewhere we can use in both places, i agree with you that if this would totally preclude a password prompt from the help text feature, we need to think hard about the next step

Comment (coryb) [inline password.go]: If help text is not directly associated with the prompt I am not sure how it could be integrated.  It is necessarily contextual to the question asked by the prompt.  FWIW, I have been doing this with some fugly in-house code (that I am trying to replace with survey) and have had very positive user feedback.

Yeah, I would do it for select and multi-select also, we would just capture the '?' key-press like we capture the arrow keys and space key (which are also inputs the user cannot see).  I have played with the implementation a bit and I think it will be okay (and intuitive).  Without setting `Help` it will be identical to the current behavior.

Comment (coryb) [inline password.go]: okay, I can revert Password.  I like that `readline` works on windows, which is why I have not tried to replace it, I didnt want to sign up to guarantee windows support on any other implementation :)  Perhaps there are simpler alternatives for readline that more closely fit what we want, or maybe I just make a readline-lite that just does the bits we like.

Comment (AlecAivazis) [inline password.go]: Yea it's probably better to just leave it up to the prompt to "decide" if it wants to support help text. 

I'm definitely interested in the feature, just want to make sure we think through all the possibilities so we don't end up having to bump a major version because we missed something obvious. I suppose it fits in the same way `Message` is defined on the prompt. Although, I might be able to see an argument to move that to the question type too, but let's not go too far down that rabbit hole just yet.

As a side note, I think it would be good to show some help text to describe the arrow key behaviour of the multiline inputs for exactly this reason - was planning on adding it soon

**Comment by coryb** [comment]
Password changes reverted.

File: renderer.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package survey
```
Comment (AlecAivazis) [inline renderer.go]: Okay, last thing before we get this in - do you mind moving this inside of `core`? I know its not very conventional, but I tend to prefer internal packages with public exports to having a large flat codebase with internal types/variables.

**Comment by AlecAivazis** [comment]
I saw, I just finished going through the code, going to take the tests for a spin and then we should be able to get this in before we go to sleep!

File: renderer.go (hunk 3 above)
Comment (coryb) [inline renderer.go]: moved it to core, had to make a bunch of stuff public since it is no longer in the same package.

Comment (AlecAivazis) [inline renderer.go]: thanks - running the last of the tests now

**Comment by coryb** [comment]
I am going to crash now, if you want more changes I will jump on it in the morning.  Thanks for spending the time on this!

**Comment by AlecAivazis** [comment]
All the tests pass except for `ask.go` and `input.go` which both need to be updated to match the new API

File: password.go (hunk 2 above)
Comment (AlecAivazis) [inline password.go]: Well ever since adding windows support to survey and getting a hang of the native windows libraries, it seemed sort of reasonable to just build a cross platform `renderer` which you sort of started already. might be worth a shot adding the necesary things to `terminal` to completely replace `readline` with `core.renderer`

File: core/renderer.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"strings"
+
+	"github.com/AlecAivazis/survey/terminal"
+	"github.com/chzyer/readline"
+)
+
+// basic readline config that does not put terminal in Raw mode
+// so that basic editing works (backspace, arrow keys etc).  This
+// is used by all text-input prompt types
+var SimpleReadlineConfig = &readline.Config{
```
Comment (AlecAivazis) [inline core/renderer.go]: Is there a reason to use this config instead of the one provided by default in `terminal.GetReadline()` which should return a configured instance that is already passed though - i figured you needed a different config here than in the default one but it's not clear to me why

Comment (coryb) [inline core/renderer.go]: Yeah, the config from `terminal.GetReadline()` still puts the tty into raw mode which prevents  the default terminal echo.  The `SimpleReadlineConfig` leaves the tty in raw mode by stubbing out the Enter/ExitRawMode callbacks.  Without the `SimpleReadlineConfig` readline does all the output control which erases my prompt via `033[J` in the runebuffer (this is the stuff I tried to work around in Password) .  The only other way to keep the readline-rawmode and get this to work is to implement an OnChange function that basically re-implements the echo functionality from your tty that is disabled by readline.  I went that way first but found that I would also have to re-implement all the native tty functionality for editing (ie backspace, arrow keys to type over content didnt work).

Comment (AlecAivazis) [inline core/renderer.go]: ahhh gotcha. thanks for the explanation - i knew you would have a good reason

**Comment by coryb** [comment]
Fixed ask.go and input.go tests to pass in new required param which is required when intializing the Prompt structs without named fields.
